### ISSUE TITLE
Single-character punctuation variables in regexps

### ISSUE THREAD

**Issue by epa** [Perl/perl5#17742]
Perl's single-character variables like `$$` `$(` `$"` and their friends let you do all sorts of powerful and concise magic.

However, in the middle of a regular expression it's easy for even experienced programmers to get in a tangle about what the escaping rules are and end up with something that's wrong, but syntactically valid because of builtin variables.  Normally to escape a special character in a regular expression you can enclose it in a character class, for example:

    /[(]/  # match literal (

But do this for the `$` character and it doesn't work:

    /[$]/  # incorrect

Moreover, the mistake is not caught at compile time.  Perl normally has good compile-time checking of regexps but here because it mentions a variable `$]` the regexp is not compiled until run time.

Could I suggest a compile time warning when one of the single-character punctuation variables apart from `$_` appears in a regular expression?  That would provide quick feedback, much quicker than waiting until the line of code is reached at run time (or waiting to run an external linter).  Personally speaking I find it rather unlikely I'd want to use `$$` `$\` `$'` and so on in a regexp, but I do get tripped up quite often by subtleties in the quoting rules.

I'm not trying to cast a moral judgement on when you might want to use these variables in a regexp, of course somebody *might*, but I think the balance tips far enough towards being a common mistake and a very rare deliberate practice that it merits a compile-time warning.  I think in the particular case of regexp syntax the programmer needs a little more hand-holding.  As far as I can tell Perl already has some special treatment of certain of these variables within regexps, with `$(` not being interpolated for example, so it wouldn't be violating purity to add a warning for the others.

**Comment by Grinnz** [comment]
I'm not going to speculate on whether it's feasible to add a useful warning or not, but this is a behavior of string interpolation in general, before it gets to the regex engine.

**Comment by epa** [comment]
> This is a behavior of string interpolation in general, before it gets to the regex engine.

I don't believe that's entirely true, there is certainly some magic going on:
````
% perl -E 'say "$$"'
2556
% perl -E '$r = qr/$$/; say $r'
(?^u:2557)
% perl -E 'say "$)"'
6508 135 1008 1701 1702 6201 6500 6508
% perl -E '$r = qr/$)/; say $r'
Unmatched ) in regex; marked by <-- HERE in m/$) <-- HERE / at -e line 1.
````
While `$$` is interpolated the same in doublequoted strings and regexps, regexp parsing has some special treatment of `$)` that doesn't interpolate it.

**Comment by Grinnz** [comment]
You're right, there are some (probably underdocumented) special cases. That one is probably because using `$` in an alternation is much more common than using `$)` in a regex.

**Comment by khwilliamson** [comment]
I looked at this a little.  The changes would have to be in toke.c
